http-client/Network/HTTP/Client/Core.hs

```diff
@@ -85,41 +86,58 @@ httpRaw'
      -> Manager
      -> IO (Request, Response BodyReader)
 httpRaw' req0 m = do
-    let req' = mSetProxy m req0
-    (req, cookie_jar') <- case cookieJar req' of
-        Just cj -> do
-            now <- getCurrentTime
-            return $ insertCookiesIntoRequest req' (evictExpiredCookies cj now) now
-        Nothing -> return (req', Data.Monoid.mempty)
-    (timeout', mconn) <- getConnectionWrapper
-        (responseTimeout' req)
-        (getConn req m)
+    ex0 <- try $ do
+        let req' = mSetProxy m req0
+        (req, cookie_jar') <- case cookieJar req' of
+            Just cj -> do
+                now <- getCurrentTime
+                return $ insertCookiesIntoRequest req' (evictExpiredCookies cj now) now
+            Nothing -> return (req', Data.Monoid.mempty)
 
-    -- Originally, we would only test for exceptions when sending the request,
-    -- not on calling @getResponse@. However, some servers seem to close
-    -- connections after accepting the request headers, so we need to check for
-    -- exceptions in both.
-    ex <- try $ do
-        cont <- requestBuilder (dropProxyAuthSecure req) (managedResource mconn)
+        (timeout', mconn) <- getConnectionWrapper
+            (responseTimeout' req)
+            (getConn req m)
 
-        getResponse timeout' req mconn cont
+        -- Originally, we would only test for exceptions when sending the request,
+        -- not on calling @getResponse@. However, some servers seem to close
+        -- connections after accepting the request headers, so we need to check for
+        -- exceptions in both.
+        ex <- try $ do
+            cont <- requestBuilder (dropProxyAuthSecure req) (managedResource mconn)
+
+            getResponse timeout' req mconn cont
+
+        case ex of
+            -- Connection was reused, and might have been closed. Try again
+            Left (e :: SomeException) -> do
+                when (managedReused mconn) $
```

Comment:
IIUC, you're treating the case where we reused a connection as a retry attempt. That would mean that code which previously had retried on a reused connection will no longer do so, because there are no retry attempts left. Is that correct?